src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -269,48 +294,99 @@ public TypeSymbolWithAnnotations SetIsAnnotated(CSharpCompilation compilation)
         /// </summary>
         internal TypeSymbolWithAnnotations MergeNullability(TypeSymbolWithAnnotations other, VarianceKind variance, out bool hadNullabilityMismatch)
         {
-            bool? isNullable = MergeIsNullable(IsNullable, other.IsNullable, variance, out bool hadTopLevelMismatch);
-            TypeSymbol type = TypeSymbol.MergeNullability(other.TypeSymbol, variance, out bool hadNestedMismatch);
+            TypeSymbol typeSymbol = other.TypeSymbol;
+            NullableAnnotation nullableAnnotation = MergeNullableAnnotation(typeSymbol, NullableAnnotation, other.NullableAnnotation, variance, out bool hadTopLevelMismatch);
+            TypeSymbol type = TypeSymbol.MergeNullability(typeSymbol, variance, out bool hadNestedMismatch);
             Debug.Assert((object)type != null);
             hadNullabilityMismatch = hadTopLevelMismatch | hadNestedMismatch;
-            return Create(type, isNullable, CustomModifiers);
+            return Create(type, nullableAnnotation, CustomModifiers);
         }
 
         /// <summary>
         /// Merges nullability.
         /// <paramref name="hadNullabilityMismatch"/> is true if there was conflict.
         /// </summary>
-        private static bool? MergeIsNullable(bool? a, bool? b, VarianceKind variance, out bool hadNullabilityMismatch)
+        private static NullableAnnotation MergeNullableAnnotation(TypeSymbol type, NullableAnnotation a, NullableAnnotation b, VarianceKind variance, out bool hadNullabilityMismatch)
         {
             hadNullabilityMismatch = false;
             if (a == b)
             {
                 return a;
             }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
             switch (variance)
             {
                 case VarianceKind.In:
-                    return (a == false || b == false) ? (bool?)false : null;
+
+                    if (a == NullableAnnotation.NotNullableBasedOnAnalysis || b == NullableAnnotation.NotNullableBasedOnAnalysis)
+                    {
+                        return NullableAnnotation.NotNullableBasedOnAnalysis;
+                    }
+
+                    if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
+                    {
+                        return NullableAnnotation.NotNullable;
+                    }
+
+                    Debug.Assert(a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown);
+                    return NullableAnnotation.Unknown;
```

Comment:
>return NullableAnnotation.Unknown; [](start = 20, length = 34)

I understand this is correct, but for readability (direct correspondence with specified behavior), consider removing the block at line 317 (`if (a.IsAnyNullable() && b.IsAnyNullable()) ...`) and adding logic here to get this structure:
 - if either `n` or `m` are non-nullable, non-nullable
 - if either `n` or `m` are oblivious, oblivious.
 - otherwise nullable.

Reply:
I intentionally moved that code before the switch in order to avoid code duplication. I believe the asserts are sufficient to help with following the logic.

---
In reply to: [231333186](https://github.com/dotnet/roslyn/pull/30913#discussion_r231333186) [](ancestors = 231333186)